### ISSUE TITLE
Alternative largest reference algorithm

### ISSUE THREAD

**Issue by martinthomson** [quicwg/base-drafts#1904]
@afrind suggested that the shorter algorithm I used for largest reference decoding might be better than the one in the doc.

This is the relevant piece of the python script used to test this (which we might also consider putting in the repo).

``` python
max_entries = int(math.floor(table_size / 32))
full_range = 2 * max_entries
if wire_lr == 0:
    lr = 0
else:
    max_value = decoder_base + max_entries
    rounded = math.floor(max_value / full_range) * full_range
    lr = rounded + wire_lr - 1
    if lr > max_value and lr >= full_range:
        lr -= full_range
```

Here until the QPACK reorg finishes.

**Comment by kazuho** [comment]
Am I correct in assuming that the following line is missing at the end?
```
    lr += 1
```

**Comment by bencebeky** [comment]
Yet another alternative is described in https://github.com/quicwg/base-drafts/pull/2248.

**Comment by afrind** [comment]
I'd like to make it a requirement that the algorithm produce an error if the encoder sent a value that is not possible if it were following the rules in the spec.  Example per the discussion in #2112:

MaxEntries = 10
TotalNumberOfInserts (decoder_base) = 2
WireLR = 18

This algorithm returns 17.  What is the guidance to convert this to an error?  Is it:

```
if lr > max_value:
  Error
```

**Comment by bencebeky** [comment]
> I'd like to make it a requirement that the algorithm produce an error if the encoder sent a value that is not possible if it were following the rules in the spec. Example per the discussion in #2112:
> 
> MaxEntries = 10
> TotalNumberOfInserts (decoder_base) = 2
> WireLR = 18
> 
> This algorithm returns 17. What is the guidance to convert this to an error? Is it:
> 
> ```
> if lr > max_value:
>   Error
> ```

I'm with you.  See issue 2249.

**Comment by martinthomson** [comment]
If you want errors on underflow:

```
    if lr > max_value and lr >= full_range:
        lr -= full_range
```
becomes
```
    if lr > max_value:
        if lr < full_range:
            throw!
        lr -= full_range
```

**Comment by afrind** [comment]
Fixed by #2112
